File: week_03/java/src/main/java/com/example/RelationKeyword.java
package com.example;

import kr.co.shineware.nlp.komoran.constant.DEFAULT_MODEL;
import kr.co.shineware.nlp.komoran.core.Komoran;
import kr.co.shineware.nlp.komoran.model.KomoranResult;

import java.io.File;
import java.io.FileInputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;

import java.nio.charset.StandardCharsets;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;

public class RelationKeyword {

    final String NEW_LINE = System.getProperty("line.separator");
    Komoran KOMORAN;

    public RelationKeyword() {

        KOMORAN = new Komoran(DEFAULT_MODEL.FULL);
        KOMORAN.setUserDic(
                "/Users/lugu/Documents/git-repositories/ds.lminho248@gmail.com/nlp_and_recommender_system/dictionary/dic.user");

    }

    // SHA-256 해싱 메소드
    public String sha256(String msg) throws NoSuchAlgorithmException {

        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(msg.getBytes());

        return bytesToHex(md.digest());

    }

    // Byte를 Hex 값으로 변환하는 메소드
    public String bytesToHex(byte[] bytes) {

        StringBuilder builder = new StringBuilder();
        for (byte b : bytes) {
            builder.append(String.format("%02x", b));
        }

        return builder.toString();

    }

    // 수집한 뉴스 데이터 세팅
    public Map<String, NewsVO> setNewsData() {

        Map<String, NewsVO> rtnMap = new LinkedHashMap<>();
        File[] newsFiles = new File(
                "/Users/lugu/Documents/git-repositories/ds.lminho248@gmail.com/nlp_and_recommender_system/news")
                .listFiles();
        Set<String> hashedSet = new HashSet<>();

        String lastField = "";
        for (File newsFile : newsFiles) {
            try {
                BufferedReader inFiles = new BufferedReader(new InputStreamReader(
                        new FileInputStream(newsFile.getAbsolutePath()), StandardCharsets.UTF_8));
                Map<String, String> item = new HashMap<>();
                String line = null;
                while ((line = inFiles.readLine()) != null) {
                    // 필드값이 포함된 라인
                    if (line.indexOf("<__") == 0 && line.indexOf("__>") > -1) {
                        String field = line.substring(3, line.indexOf("__>"));
                        String value = line.substring(line.indexOf("__>") + 3);

                        // Map에 키가 중복될 경우, 뉴스 세팅 완료
                        if (item.containsKey(field)) {
                            // 세팅하기 전에 해시값이 중복되었는지 체크 (뉴스 중복 체크)
                            String hashedText = sha256(item.get("url"));
                            if (!hashedSet.contains(hashedText)) {
                                // Map to NewVO를 한 후, 리스트에 세팅 (section_id)를 pk로 설정
                                rtnMap.put(item.get("section") + "_" + item.get("id"), mapToNewsVO(item));
                            }

                            hashedSet.add(hashedText);
                            // Map 초기화
                            item = new HashMap<>();
                        }

                        item.put(field, value);
                        lastField = field;
                    } else {
                        // 필드값을 못 찾는 경우 (value만 존재)
                        StringBuffer sb = new StringBuffer();
                        sb.append(item.get(lastField) + NEW_LINE);
                        sb.append(line);
                        item.put(lastField, sb.toString());
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        return rtnMap;
    }

    public NewsVO mapToNewsVO(Map<String, String> item) {

        NewsVO vo = new NewsVO();

        vo.setId(item.get("id"));
        vo.setUrl(item.get("url"));
        vo.setSection(item.get("section"));
        vo.setSubSection(item.get("subSection"));
        vo.setTitle(item.get("title"));
        vo.setSummary(item.get("summary"));
        vo.setDttm(item.get("dttm"));
        vo.setMedia(item.get("media"));
        vo.setContent(item.get("content"));

        return vo;

    }

    public Map<String, Map<String, List<String>>> extractNounMap(Map<String, NewsVO> newsMap) {

        Map<String, Map<String, List<String>>> rtnMap = new HashMap<>();
        for (String pk : newsMap.keySet()) {
            Map<String, List<String>> item = new HashMap<>();

            NewsVO vo = newsMap.get(pk);

            item.put("title", extractNouns(vo.getTitle()));
            item.put("summary", extractNouns(vo.getSummary()));
            item.put("content", extractNouns(vo.getContent()));

            rtnMap.put(pk, item);
        }

        return rtnMap;

    }

    public List<String> extractNouns(String document) {

        List<String> nounList = new ArrayList<>();
        if (document.trim().length() == 0)
            return nounList;

        KomoranResult analyzeResultList = KOMORAN.analyze(document);
        nounList = analyzeResultList.getNouns();

        return nounList;

    }

    public Map<String, Map<String, Integer>> calTF(Map<String, Map<String, List<String>>> nounsMap,
            Map<String, NewsVO> newsMap) {

        Map<String, Map<String, Integer>> sectKwdCntMap = new HashMap<>();
        for (String pk : nounsMap.keySet()) {
            NewsVO vo = newsMap.get(pk);

            // 이전 섹션값별 키워드 카운트 값을 가져온다.
            Map<String, Integer> kwdCntMap = new HashMap<>();
            if (sectKwdCntMap.containsKey(vo.getSection())) {
                kwdCntMap = sectKwdCntMap.get(vo.getSection());
            }

            // pk별 nouns 데이터를 읽은 후 카운트 세팅
            Map<String, List<String>> itemNounMap = nounsMap.get(pk);
            for (String field : itemNounMap.keySet()) {
                List<String> nounList = itemNounMap.get(field);

                for (String noun : nounList) {
                    int cnt = 1;

                    if (kwdCntMap.containsKey(noun)) {
                        cnt += kwdCntMap.get(noun);
                    }
                    kwdCntMap.put(noun, cnt);
                }
            }

            sectKwdCntMap.put(vo.getSection(), kwdCntMap);
        }

        return sectKwdCntMap;

    }

    // 카운트 순으로 정렬
    public Map<String, Set<String>> sortedKwdMap(Map<String, Map<String, Integer>> sectKwdCntMap, int limit) {

        Map<String, Set<String>> rtnMap = new HashMap<>();
        for (String sect : sectKwdCntMap.keySet()) {
            Set<String> corpus = new LinkedHashSet<>();
            Map<String, Integer> kwdCntMap = sectKwdCntMap.get(sect);

            // 정렬
            Map<String, Integer> result = kwdCntMap.entrySet().stream()
                    .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> oldValue,
                            LinkedHashMap::new));

            for (String kwd : result.keySet()) {
                corpus.add(kwd);
                if (corpus.size() >= limit)
                    break;
            }

            rtnMap.put(sect, corpus);
        }

        return rtnMap;
    }

    // 연관 단어 생성
    public Map<String, List<String>> relationKwds(Map<String, NewsVO> newsMap,
            Map<String, Map<String, List<String>>> nounsMap, Map<String, Set<String>> sectCorpus) {

        Map<String, Map<String, Integer>> relationCntMap = new HashMap<>();

        for (String pk : nounsMap.keySet()) {
            Map<String, List<String>> fieldKwdList = nounsMap.get(pk);
            Set<String> corpus = sectCorpus.get(newsMap.get(pk).getSection());

            // 한 문서에 같이 등장하면 연관성이 있는 단어로 인정
            Set<String> docKwdSet = new HashSet<>();
            for (String field : fieldKwdList.keySet()) {
                docKwdSet.addAll(fieldKwdList.get(field));
            }

            // 생성된 docKwdSet을 단어별 연관 키워드로 생성
            for (String kwd : docKwdSet) {
                Map<String, Integer> kwdCntMap = new HashMap<>();

                if (relationCntMap.containsKey(kwd))
                    kwdCntMap = relationCntMap.get(kwd);

                for (String otherKwd : docKwdSet) {
                    if (kwd.equals(otherKwd) || !corpus.contains(otherKwd) || otherKwd.length() <= 1)
                        continue;

                    int cnt = 1;
                    if (kwdCntMap.containsKey(otherKwd)) {
                        cnt += kwdCntMap.get(otherKwd);
                    }

                    kwdCntMap.put(otherKwd, cnt);
                }

                relationCntMap.put(kwd, kwdCntMap);
            }
        }

        return calcRelationKwds(relationCntMap);

    }

    // 카운트 기반 내림차순 정렬 후 상위 연관 키워드를 추출
    public Map<String, List<String>> calcRelationKwds(Map<String, Map<String, Integer>> relationCntMap) {

        Map<String, List<String>> relationKwdMap = new HashMap<>();

        for (String kwd : relationCntMap.keySet()) {
            Map<String, Integer> kwdCntMap = relationCntMap.get(kwd);

            // 정렬
            Map<String, Integer> result = kwdCntMap.entrySet().stream()
                    .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                    .collect(Collectors.toMap(
                            Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> oldValue,
                            LinkedHashMap::new));

            List<String> sortedKwdList = new ArrayList<>();
            for (String sortedKwd : result.keySet()) {
                sortedKwdList.add(sortedKwd);
                if (sortedKwdList.size() >= 10)
                    break;
            }

            relationKwdMap.put(kwd, sortedKwdList);
        }

        return relationKwdMap;

    }

    // 연관 키워드 출력
    public void inputKwd(Map<String, List<String>> relationKwdMap) {

        while (true) {
            Scanner sc = new Scanner(System.in);
            System.out.print("키워드를 입력하세요. : ");
            String ques = sc.nextLine();

            if (ques.trim().equals("끝"))
                System.exit(1);

            if (relationKwdMap.containsKey(ques.trim())) {
                System.out.println(relationKwdMap.get(ques.trim()));
            } else {
                System.out.println("연관 키워드가 존재하지 않습니다.");
            }
        }

    }

    public void execute() {

        Map<String, NewsVO> newsMap = setNewsData();
        // for (String pk : newsMap.keySet()) {
        // System.out.println(pk + "->>" + newsMap.get(pk));
        // System.exit(1);
        // }

        // 컬럼별 형태소 분석 수행 (pk, 필드, 명사리스트)
        Map<String, Map<String, List<String>>> nounsMap = extractNounMap(newsMap);
        // for (String pk : nounsMap.keySet()) {
        // System.out.println(pk + " -> " + nounsMap.get(pk));
        // System.exit(1);
        // }

        // 뉴스 섹션별(6종) TF 기반 카운트 생성 (파일 저장 및 TF-Corpus 메소드, 20분?)
        Map<String, Map<String, Integer>> sectKwdCntMap = calTF(nounsMap, newsMap);
        // for (String section : sectKwdCntMap.keySet()) {
        // Map<String, Integer> kwdCntMap = sectKwdCntMap.get(section);

        // int count = 0;
        // for (String kwd : kwdCntMap.keySet()) {
        // System.out.println(kwd + " -> " + kwdCntMap.get(kwd));
        // count++;

        // if (count > 100)
        // break;
        // }
        // System.exit(1);
        // }

        // sectKwdCntMap을 내림차순 정렬 후 상위값 가져오기
        Map<String, Set<String>> sectCorpus = sortedKwdMap(sectKwdCntMap, 500);

        // 연관 단어 생성
        Map<String, List<String>> relationKwdMap = relationKwds(newsMap, nounsMap, sectCorpus);

        inputKwd(relationKwdMap);

    }
}

File: week_03/python/relation_keyword.py
import os
import re
import sys

from tqdm import tqdm

from collections import defaultdict, OrderedDict

from konlpy.tag import Komoran
from konlpy.utils import pprint


CUR_DIR = os.path.dirname(os.path.realpath(__file__))
PAR_DIR = os.path.dirname(os.path.dirname(CUR_DIR))
NEWS_DATA_PATH = os.path.join(PAR_DIR, 'news')
USER_DIC = os.path.join(PAR_DIR, 'dictionary', 'dic.user')


class RelationKeyword:
    
    def __init__(self):
        
        self.file_list = [
            os.path.join(NEWS_DATA_PATH, f) \
            for f in os.listdir(NEWS_DATA_PATH) \
            if f.split('.')[-1] == 'txt'
        ]
        
        self.new_line = "\n"
        self.komoran = Komoran(userdic=USER_DIC)
    
    def _set_news_data(self):
        
        field_pattern = re.compile('<__([a-z\_]+)__>')
        
        lastField = ""
        rtnMap = defaultdict(dict)
        for f in self.file_list:
            o = open(f, 'r', encoding='UTF-8')
            lines = o.readlines()
            
            item = OrderedDict()
            for line in lines:
                if re.search(field_pattern, line) is not None:
                    field = re.search(field_pattern, line).groups()[0].strip()
                    field_end_point = re.search(field_pattern, line).end()
                    value = line[field_end_point:]

                    if field in item.keys():
                        rtnMap[item.get('section') + "_" + item.get('id')] = item
                        item = OrderedDict()
                    
                    item[field] = value.rstrip()
                    lastField = field
                else:
                    sb = item.get(lastField) + line
                    item[lastField] = sb
            else:
                rtnMap[item.get('section') + "_" + item.get('id')] = item
            
        return rtnMap
    
    def _extract_noun_map(self, newsMap:dict):
        
        rtnMap = defaultdict(dict)
        
        for pk in newsMap.keys():
            item = OrderedDict()
            vo = newsMap[pk]
            
            item['title'] = self._extract_nouns(vo.get('title'))
            item['summary'] = self._extract_nouns(vo.get('summary'))
            item['content'] = self._extract_nouns(vo.get('content'))
            
            rtnMap[pk] = item
        
        return rtnMap
    
    def _extract_nouns(self, doc:str):
        
        nounList = []
        if doc.strip() == "": return nounList
        
        nounList = self.komoran.nouns(doc)
        
        return nounList
    
    def _cal_TF(self, nounsMap:dict, newsMap:dict):
        
        sectKwdCntMap = defaultdict(dict)
        for pk in nounsMap.keys():
            vo = newsMap.get(pk)
            
            kwdCntMap = OrderedDict()
            if vo['section'] in sectKwdCntMap.keys():
                kwdCntMap = sectKwdCntMap.get(vo['section'])
            
            itemNounMap = nounsMap.get(pk)
            for field in itemNounMap.keys():
                nounList = itemNounMap.get(field)
                
                for noun in nounList:
                    if noun in kwdCntMap.keys():
                        kwdCntMap[noun] += 1
                    else:
                        kwdCntMap[noun] = 1
            
            sectKwdCntMap[vo['section']] = kwdCntMap
        
        return sectKwdCntMap
    
    def _sorted_kwd_map(self, sectKwdCntMap:dict, limit:int=500):
        
        rtnMap = defaultdict(set)
        for sect in sectKwdCntMap.keys():
            corpus = set()
            kwdCntMap = sectKwdCntMap.get(sect)
            sortedKwdCntMap = dict(sorted(kwdCntMap.items(), key=lambda x: x[1], reverse=True))
            
            for kwd in sortedKwdCntMap.keys():
                if len(kwd.strip()) < 2: continue
                corpus.add(kwd)
                if len(corpus) >= limit: break
            
            rtnMap[sect] = corpus
            self._corpus_to_file(corpus, out_fn=os.path.join(PAR_DIR, 'corpus', f'{sect}.tsv'))
        
        return rtnMap
    
    def _corpus_to_file(self, corpus:set, out_fn:str=None):
        
        import csv
        with open(out_fn, 'w', newline='') as f:
            writer = csv.writer(f, delimiter='\t')
            writer.writerows(corpus)
        
        return None
    
    def _relation_kwds(self, newsMap:dict, nounsMap:dict, sectCorpus:dict):
        
        relationCntMap = defaultdict(dict)
        for pk in nounsMap.keys():
            fieldKwdList = nounsMap.get(pk)
            corpus = sectCorpus.get(newsMap.get(pk)['section'])
            
            docKwdSet = set()
            for field in fieldKwdList.keys():
                docKwdSet |= set(fieldKwdList.get(field))
            
            for kwd in docKwdSet:
                kwdCntMap = OrderedDict()
                if kwd in relationCntMap.keys():
                    kwdCntMap = relationCntMap.get(kwd)
                
                for otherKwd in docKwdSet:
                    if (kwd == otherKwd) or (otherKwd not in corpus) or (len(otherKwd) < 2):
                        continue
                    
                    if otherKwd in kwdCntMap.keys():
                        kwdCntMap[otherKwd] += 1
                    else:
                        kwdCntMap[otherKwd] = 1
                
                relationCntMap[kwd] = kwdCntMap
        
        return self._calc_relation_kwds(relationCntMap)
    
    def _calc_relation_kwds(self, relationCntMap:dict, limit:int=10):
        
        relationKwdMap = defaultdict(list)
        for kwd in relationCntMap.keys():
            kwdCntMap = relationCntMap.get(kwd)
            sortedKwdCntMap = dict(sorted(kwdCntMap.items(), key=lambda x: x[1], reverse=True))
            
            sortedKwdList = []
            for sortedKwd in sortedKwdCntMap.keys():
                sortedKwdList.append(sortedKwd)
                if len(sortedKwdList) >= limit: break
            
            relationKwdMap[kwd] = sortedKwdList
        
        return relationKwdMap
    
    def _input_kwd(self, relationKwdMap):
        
        while True:
            userInput = input("키워드를 입력하세요 : ")
            
            if userInput.strip() == "끝": sys.exit(1)
            
            if userInput.strip() in relationKwdMap.keys():
                print(relationKwdMap.get(userInput.strip()))
            else:
                print("연관 키워드가 존재하지 않습니다.")
        
        return None
    
    def execute(self):
        
        newsMap = self._set_news_data()
        nounsMap = self._extract_noun_map(newsMap)
        sectKwdCntMap = self._cal_TF(nounsMap, newsMap)
        sectCorpus = self._sorted_kwd_map(sectKwdCntMap)
        relationKwdMap = self._relation_kwds(newsMap, nounsMap, sectCorpus)
        self._input_kwd(relationKwdMap)
        
        return None